### PR TITLE
LR2 - Service Station by Mariya Protskaya

### PR DESCRIPTION
**Pull request by larielrary**

### CODE REVIEW

File: ServiceStation/BusinessLayer/Models/Car.cs [hunk 1]
```diff
@@ -0,0 +1,30 @@
+﻿using DataLayer.Entity;
+using System.ComponentModel.DataAnnotations;
+using System.Runtime.Serialization;
+
+namespace BusinessLayer.Models.DTO
+{
+    [DataContract]
+    public class Car : IEntity
+    {
+        [Key]
```
Comment (alexey62soft) [inline ServiceStation/BusinessLayer/Models/Car.cs]: DAL attribute should absent in BLL models.

File: ServiceStation/BusinessLayer/ServiceStationService/CarService.cs [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+            if (car == null)
+            {
+                throw new ArgumentNullException(nameof(car));
+            }
+            else
+            {
+                await _carRepository.Delete(_mapper.Map<CarDTO>(car[0]).Id);
+            }
+        }
+        public async Task Update(Car car)
```
Comment (alexey62soft) [inline ServiceStation/BusinessLayer/ServiceStationService/CarService.cs:50]: All separate lines between all methods.

File: ServiceStation/BusinessLayer/ServiceStationService/CarTechnicalConditionService.cs [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+        public async Task<CarTechnicalCondition> GetItem(int id)
+        {
+            var condition = await _conditionRepository.GetById(id);
+
+            if (condition == null)
+            {
+                throw new ArgumentNullException(nameof(condition));
+            }
+
+            return new CarTechnicalCondition
```
Comment (alexey62soft) [inline ServiceStation/BusinessLayer/ServiceStationService/CarTechnicalConditionService.cs]: Automapper?

File: ServiceStation/BusinessLayer/ServiceStationService/InspectorService.cs [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+        public async Task<Inspector> GetItem(int id)
+        {
+            var inspector = await _inspectorRepository.GetById(id);
+
+            if (inspector == null)
+            {
+                throw new ArgumentNullException(nameof(inspector));
+            }
+
+            return new Inspector
```
Comment (alexey62soft) [inline ServiceStation/BusinessLayer/ServiceStationService/InspectorService.cs]: Automapper?

File: ServiceStation/BusinessLayer/ServiceStationService/CarService.cs [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+        }
+        public async Task<Car> GetItem(int id)
+        {
+            var car = await _carRepository.GetById(id);
+            if (car == null)
+            {
+                throw new ArgumentNullException(nameof(car));
+            }
+
+            return new Car
```
Comment (alexey62soft) [inline ServiceStation/BusinessLayer/ServiceStationService/CarService.cs]: Automapper?

File: ServiceStation/BusinessLayer/ServiceStationService/OwnerService.cs [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+        public async Task<Owner> GetItem(int id)
+        {
+            var owner = await _ownerRepository.GetById(id);
+
+            if (owner == null)
+            {
+                throw new ArgumentNullException(nameof(owner));
+            }
+
+            return new Owner
```
Comment (alexey62soft) [inline ServiceStation/BusinessLayer/ServiceStationService/OwnerService.cs]: Automapper?

File: ServiceStation/BusinessLayer/ServiceStationService/CarService.cs [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+        public async Task<IEnumerable<Car>> GetItems()
+        {
+            return _mapper.Map<IEnumerable<CarDTO>, List<Car>>(await _carRepository.GetAll());
+        }
+        public async Task<Car> GetItem(int id)
+        {
+            var car = await _carRepository.GetById(id);
+            if (car == null)
+            {
+                throw new ArgumentNullException(nameof(car));
```
Comment (alexey62soft) [inline ServiceStation/BusinessLayer/ServiceStationService/CarService.cs]: It should be InvalidOperationException.
Fix in all services.

File: ServiceStation/BusinessLayer/ServiceStationService/CarService.cs [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+                else
+                {
+                    await _carRepository.Add(_mapper.Map<CarDTO>(car));
+                }
+            }
+        }
+
+        public async Task Delete(int id)
+        {
+            var car = GetItems().Result.Where(el => el.Id == id).ToList();
```
Comment (alexey62soft) [inline ServiceStation/BusinessLayer/ServiceStationService/CarService.cs]: (await GetItems()).Single()

Fix in all places

File: ServiceStation/BusinessLayer/ServiceStationService/CarService.cs [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+            if (car == null)
```
Comment (alexey62soft) [inline ServiceStation/BusinessLayer/ServiceStationService/CarService.cs]: In case of using construction above - this condition useless.

Comment (alexey62soft) [reply ServiceStation/BusinessLayer/ServiceStationService/CarService.cs]: Fix in all services

File: ServiceStation/BusinessLayer/BusinessLayer.csproj [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+  </PropertyGroup>
+
+  <ItemGroup>
+    <PackageReference Include="AutoMapper" Version="9.0.0" />
+    <PackageReference Include="Microsoft.EntityFrameworkCore" Version="3.1.3" />
+    <PackageReference Include="Microsoft.Extensions.Configuration.FileExtensions" Version="3.1.4" />
+    <PackageReference Include="Microsoft.Extensions.Configuration.Json" Version="3.1.4" />
+    <PackageReference Include="Microsoft.Extensions.DependencyInjection" Version="3.1.3" />
+    <PackageReference Include="Ninject" Version="3.3.4" />
+    <PackageReference Include="NLog.Extensions.Logging" Version="1.6.4" />
```
Comment (alexey62soft) [inline ServiceStation/BusinessLayer/BusinessLayer.csproj]: All fo them should be removed here.

File: ServiceStation/ConsoleApp/Menu.cs [hunk 11]
```diff
@@ -0,0 +1,392 @@
...
+        public Menu(CarService car, CarTechnicalConditionService carCondition,
+                                OwnerService owner, InspectorService inspector)
+        {
+            this.car = car;
+            this.carCondition = carCondition;
+            this.owner = owner;
+            this.inspector = inspector;
+        }
+
+        public async Task Start()
```
Comment (alexey62soft) [inline ServiceStation/ConsoleApp/Menu.cs]: Split menu by entities.
it's very hard to review.

File: ServiceStation/ConsoleApp/Runner.cs [hunk 12]
```diff
@@ -0,0 +1,18 @@
+﻿using Microsoft.Extensions.Logging;
+
+namespace ConsoleApp
+{
+    public class Runner
```
Comment (alexey62soft) [inline ServiceStation/ConsoleApp/Runner.cs]: Why do you need this?)

File: ServiceStation/DataLayer/DataLayer.csproj [hunk 13]
```diff
@@ -0,0 +1,15 @@
...
+  <PropertyGroup>
+    <TargetFramework>netstandard2.0</TargetFramework>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <PackageReference Include="Microsoft.EntityFrameworkCore" Version="3.1.3" />
+    <PackageReference Include="Microsoft.Extensions.Configuration.FileExtensions" Version="3.1.4" />
+    <PackageReference Include="Microsoft.Extensions.Configuration.Json" Version="3.1.4" />
+    <PackageReference Include="Microsoft.Extensions.DependencyInjection" Version="3.1.3" />
+    <PackageReference Include="NLog.Extensions.Logging" Version="1.6.4" />
```
Comment (alexey62soft) [inline ServiceStation/DataLayer/DataLayer.csproj]: All of them should be removed.

File: ServiceStation/DataLayer/Models/CarDTO.cs [hunk 14]
```diff
@@ -0,0 +1,15 @@
+﻿using DataLayer.Entity;
+
+namespace DataLayer.Models
+{
+    public class CarDTO:IEntity
```
Comment (alexey62soft) [inline ServiceStation/DataLayer/Models/CarDTO.cs]: Fix formatting.
It should be consistent across all entities.

File: ServiceStation/DataLayer/Repository/ServiceStationRepository.cs [hunk 15]
```diff
@@ -0,0 +1,51 @@
+﻿using DataLayer.Entity;
+using Microsoft.EntityFrameworkCore;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+
+namespace DataLayer.Repository
+{
+    public class ServiceStationRepository<T> : IRepository<T> where T : class, IEntity
```
Comment (alexey62soft) [inline ServiceStation/DataLayer/Repository/ServiceStationRepository.cs]: `GenericRepository<T>`

**Review by alexey62soft — CHANGES_REQUESTED** [review]

**Comment by alexey62soft** [comment]
Also, the report should be uploaded.

File: ServiceStation/ConsoleApp/Runner.cs (hunk 12 above)
Comment (larielrary) [reply ServiceStation/ConsoleApp/Runner.cs]: For a service created by a dependency injection (AddTransient method in Startup)

**Comment by larielrary** [comment]
All issues fixed.

File: ServiceStation/ConsoleApp/Runner.cs (hunk 12 above)
Comment (alexey62soft) [reply ServiceStation/ConsoleApp/Runner.cs]: Where this service used?)

Comment (alexey62soft) [reply ServiceStation/ConsoleApp/Runner.cs]: I see only registration, but not usage.

Comment (larielrary) [reply ServiceStation/ConsoleApp/Runner.cs]: Startup class, line 22

File: ServiceStation/ConsoleApp/Menu.cs (hunk 11 above)
Comment (alexey62soft) [reply ServiceStation/ConsoleApp/Menu.cs]: Hm.
You not understand me)
You should split one menu class into few menus classe for existing entities.
Each menu class will be responsible only for the one entity.
For example, StationsMenu class will provide menu for stations and all operations will be in it.

**Comment by alexey62soft** [comment]
I think two comments still not resolved.
Also, formatting should be consistent across all solution.
Please, check whole solution and fix all inconsistent moments.

**Comment by alexey62soft** [comment]
ServiceStationsRepository should be removed.

File: ServiceStation/ConsoleApp/Menu.cs (hunk 11 above)
Comment (larielrary) [reply ServiceStation/ConsoleApp/Menu.cs]: Fixed

File: ServiceStation/ConsoleApp/Runner.cs (hunk 12 above)
Comment (alexey62soft) [reply ServiceStation/ConsoleApp/Runner.cs]: Its a registration, but where the usage?)))

**Comment by alexey62soft** [comment]
I guess the interface for menu should be created with the methods:
CreateAsync
EditAsync
DeleteAsync
All menu's methods should be not static.

All menus should be registered in DI (can be a singleton) and fetched in root Menu constructor.

**Comment by alexey62soft** [comment]
https://github.com/larielrary/ProtskayaMariya_dotNet/pull/2#discussion_r429576818
Still not resolved)

**Comment by larielrary** [comment]
> I guess the interface for menu should be created with the methods:
> CreateAsync
> EditAsync
> DeleteAsync
> All menu's methods should be not static.
> 
> All menus should be registered in DI (can be a singleton) and fetched in root Menu constructor.

fixed

File: ServiceStation/ConsoleApp/ConsoleApp.csproj [hunk 16]
```diff
@@ -0,0 +1,28 @@
...
+
+  <ItemGroup>
+    <PackageReference Include="AutoMapper.Extensions.Microsoft.DependencyInjection" Version="7.0.0" />
+    <PackageReference Include="Microsoft.AspNetCore.Identity" Version="2.2.0" />
+    <PackageReference Include="Microsoft.AspNetCore.Identity.EntityFrameworkCore" Version="3.1.3" />
+    <PackageReference Include="Microsoft.Extensions.Configuration" Version="3.1.4" />
+    <PackageReference Include="Microsoft.Extensions.DependencyInjection" Version="3.1.4" />
+    <PackageReference Include="NLog" Version="4.7.2" />
+    <PackageReference Include="System.Configuration.ConfigurationManager" Version="4.7.0" />
+    <PackageReference Include="Microsoft.EntityFrameworkCore.SqlServer" Version="3.1.4" />
```
Comment (alexey62soft) [inline ServiceStation/ConsoleApp/ConsoleApp.csproj]: This lib should be in DAL

File: ServiceStation/ConsoleApp/ConsoleApp.csproj [hunk 17]
```diff
@@ -0,0 +1,28 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <OutputType>Exe</OutputType>
+    <TargetFramework>netcoreapp3.1</TargetFramework>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <PackageReference Include="AutoMapper.Extensions.Microsoft.DependencyInjection" Version="7.0.0" />
+    <PackageReference Include="Microsoft.AspNetCore.Identity" Version="2.2.0" />
```
Comment (alexey62soft) [inline ServiceStation/ConsoleApp/ConsoleApp.csproj]: Libs on line 10-11 are useless and should be removed

**Review by alexey62soft — CHANGES_REQUESTED** [review]

File: ServiceStation/ConsoleApp/ConsoleApp.csproj (hunk 17 above)
Comment (larielrary) [reply ServiceStation/ConsoleApp/ConsoleApp.csproj]: Fixed

File: ServiceStation/ConsoleApp/ConsoleApp.csproj (hunk 16 above)
Comment (larielrary) [reply ServiceStation/ConsoleApp/ConsoleApp.csproj]: Fixed

**Review by alexey62soft — APPROVED** [review]

**Comment by alexey62soft** [comment]
Upload the report, please.

**Comment by larielrary** [comment]
> Upload the report, please.


Report was uploaded a week ago)

**Pull request merged**